frontend/src/sagas.js

```diff
@@ -794,110 +817,108 @@ export function* fetchAllProjectDetailsSaga({ projectId, showLoading }) {
   }, showLoading);
 }
 
-export function* fetchFirstProjectHistorySaga({ projectId, offset, limit, showLoading }) {
+export function* fetchNextProjectHistoryPageSaga({ projectId, showLoading }) {
   yield execute(function*() {
-    if (limit <= 0) {
-      return;
+    const { currentHistoryPage, historyPageSize, totalHistoryItemCount } = yield select(getProjectHistoryState);
+
+    let offset = 0;
+    if (totalHistoryItemCount === 0) {
+      // Before the first call, we don't know how many history items there are
+      // so we fetch a fixed number of the latest items
+      offset = -historyPageSize;
+    } else {
+      // After the first call we just fetch each page
+      // If the offset is below 0, we have reached the last page
+      // and can fetch the first events
+      offset = totalHistoryItemCount - (currentHistoryPage + 1) * historyPageSize;
+      offset = offset > 0 ? offset : 0;
     }
-    const { historyItemsCount, events } = yield callApi(api.viewProjectHistory, projectId, offset, limit);
-    const newOffset = historyItemsCount + offset - limit;
-    offset = newOffset > 0 ? newOffset : 0;
-    const hasMore = offset + limit < historyItemsCount;
-    yield put({
-      type: FETCH_FIRST_PROJECT_HISTORY_SUCCESS,
-      offset,
-      limit,
-      historyItemsCount,
-      events,
-      hasMore
-    });
-  }, showLoading);
-}
 
-export function* fetchNextProjectHistorySaga({ projectId, offset, limit, showLoading }) {
-  yield execute(function*() {
-    if (limit <= 0) {
-      return;
+    // If the offset is 0, we are on the last page and only need to fetch the remaining items
+    const limit = offset === 0 ? totalHistoryItemCount % historyPageSize : historyPageSize;
```

Comment:
Do we really need to do this? I believe the API anyways only returns the amount of items it has

Reply:
`Math.max(0, totalHistoryItemCount - (historyPageSize * currentHistoryPage))` won't work because the limit has to be greater than 0 (restriction from the API). This means that we cannot use `limit = 0` for the case where we expect the API not to return anything.

I tested the case where the number of history items is a multiple of the page size (100 items and pagesize of 50) with this code 
```js
const limit =
      offset === 0 && totalHistoryItemCount % historyPageSize !== 0
        ? totalHistoryItemCount % historyPageSize
        : historyPageSize
```
and it works as expected. We could also use 
```js
    const limit =
      offset === 0 && totalHistoryItemCount % historyPageSize !== 0
        ? totalHistoryItemCount - historyPageSize * currentHistoryPage
        : historyPageSize;
```
which is basically the same.

Reply:
suggestion to make the code clearer and easier to read: 
```js
    const isLastPage = offset === 0;
    const remainingItems = totalHistoryItemCount - currentHistoryPage * historyPageSize;
    const limit = isLastPage && remainingItems !== 0 ? remainingItems : historyPageSize
```